File: Delphi/DBComparerDemoUniDAC/Unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, Buttons, StdCtrls, ExtCtrls, dbcDBComparer, dbcDBStructure,
  dbcIBDatabaseExtract, dbcClasses, dbcIBScriptExtract, dbcSQL_Exec, dbcIBSQLExec,
  ComCtrls, dbcDBEngine, dbcCustomScriptExtract,
  dbcTableDataComparer;

type
  TForm1 = class(TForm)
    IBDBExtract1: TIBDBExtract;
    DBStructure1: TDBStructure;
    DBStructure2: TDBStructure;
    IBDBExtract2: TIBDBExtract;
    OpenDialog1: TOpenDialog;
    GroupBox1: TGroupBox;
    fnDatabaseMaster: TEdit;
    fnDatabaseTarget: TEdit;
    Label1: TLabel;
    Label2: TLabel;
    BtnSelDatabaseMaster: TSpeedButton;
    BtnSelDatabaseTarget: TSpeedButton;
    EditMUser: TEdit;
    EditMPswd: TEdit;
    LabelMUser: TLabel;
    LabelMPswd: TLabel;
    LabelTUser: TLabel;
    EditTUser: TEdit;
    LabelTPswd: TLabel;
    EditTPswd: TEdit;
    BtnExtract: TBitBtn;
    BtnCompare: TBitBtn;
    GroupBox2: TGroupBox;
    MemoExtr: TRichEdit;
    DBComparer1: TDBComparer;
    Splitter1: TSplitter;
    GroupBox3: TGroupBox;
    MemoResult: TMemo;
    fnScriptMaster: TEdit;
    BtnSelScriptMaster: TSpeedButton;
    fnScriptTarget: TEdit;
    BtnSelScriptTarget: TSpeedButton;
    IBScriptExtract1: TIBScriptExtract;
    IBScriptExtract2: TIBScriptExtract;
    Panel1: TPanel;
    rbMasterFromDatabase: TRadioButton;
    rbMasterFromScript: TRadioButton;
    Panel2: TPanel;
    rbTargetFromDatabase: TRadioButton;
    rbTargetFromScript: TRadioButton;
    IBSQLExec: TIBSQLExec;
    BtnUpdate: TBitBtn;
    Gauge1: TProgressBar;
    Label3: TLabel;
    GroupBox4: TGroupBox;
    EditTableM: TEdit;
    EditTableT: TEdit;
    Label4: TLabel;
    Label5: TLabel;
    TableDataComparer: TTableDataComparer;
    Label6: TLabel;
    cbCompareMode: TComboBox;
    EditCustomFlds: TEdit;
    LabelCustomFlds: TLabel;
    EditExclFlds: TEdit;
    Label8: TLabel;
    BtnDataCompare: TBitBtn;
    BitBtn2: TBitBtn;
    Label9: TLabel;
    Gauge2: TProgressBar;
    procedure FormCreate(Sender: TObject);
    procedure BtnSelFileClick(Sender: TObject);
    procedure BtnExtractClick(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure LogNextLine(Sender: TObject; LogLine: String);
    procedure BtnCompareClick(Sender: TObject);
    procedure rbFromClick(Sender: TObject);
    procedure ProgressUpdate(Sender: TObject; Value, MaxValue: Integer);
    procedure BtnUpdateClick(Sender: TObject);
    procedure BtnDataCompareClick(Sender: TObject);
    procedure EditTableMChange(Sender: TObject);
    procedure TableDataComparerProgressUpdateMaster(Sender: TObject; Value,
      MaxValue: Integer);
    procedure TableDataComparerProgressUpdateTarget(Sender: TObject; Value,
      MaxValue: Integer);
    procedure cbCompareModeChange(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
    procedure DecodeFields(aFlds: String; aFields: TStrings);
    procedure ExtractMasterDatabase();
    procedure ExtractTargetDatabase();
    procedure ExtractMasterScript();
    procedure ExtractTargetScript();
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

Uses IBDataModule, IniFiles, dbcCompObj;

function IniFileName(): String;
begin
  Result := ExtractFilePath(ParamStr(0))+'CompDemo.Ini';
end;

procedure AddToExtrlog(s: String);
begin
  Form1.MemoExtr.Lines.Add(s);
end;

procedure TForm1.FormCreate(Sender: TObject);
var
  MFromD, TFromD: Boolean;
begin
  fnDatabaseMaster.Text := '';
  fnDatabaseTarget.Text := '';
  fnScriptMaster.Text := '';
  fnScriptTarget.Text := '';
  MemoExtr.Clear;
  MemoResult.Clear;
  with TIniFile.Create(IniFileName()) do begin
    fnDatabaseMaster.Text := ReadString('Databases','MasterDB','');
    fnDatabaseTarget.Text := ReadString('Databases','TargetDB','');
    fnScriptMaster.Text := ReadString('Databases','MasterScript','');
    fnScriptTarget.Text := ReadString('Databases','TargetScript','');
    MFromD := ReadBool('Databases','MasterFromDatabase',True);
    TFromD := ReadBool('Databases','TargetFromDatabase',True);
    if MFromD then rbMasterFromDatabase.Checked := True
    else rbMasterFromScript.Checked := True;
    if TFromD then rbTargetFromDatabase.Checked := True
    else rbTargetFromScript.Checked := True;
    EditTableM.Text := ReadString('TableDataComparer','TableNameMaster', '');
    EditTableT.Text := ReadString('TableDataComparer','TableNameTarget', '');
    cbCompareMode.ItemIndex := ReadInteger('TableDataComparer','DataCompareMode', 0);
    cbCompareModeChange(cbCompareMode);
    EditCustomFlds.Text := ReadString('TableDataComparer','CustomKeyFields', '');
    EditExclFlds.Text := ReadString('TableDataComparer','ExcludeDataFields', '');
    Free;
  end;
  rbFromClick(nil);
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
  with TIniFile.Create(IniFileName()) do begin
    WriteString('Databases', 'MasterDB', fnDatabaseMaster.Text);
    WriteString('Databases', 'TargetDB', fnDatabaseTarget.Text);
    WriteString('Databases', 'MasterScript', fnScriptMaster.Text);
    WriteString('Databases', 'TargetScript', fnScriptTarget.Text);
    WriteBool('Databases','MasterFromDatabase',rbMasterFromDatabase.Checked);
    WriteBool('Databases','TargetFromDatabase',rbTargetFromDatabase.Checked);
    WriteString('TableDataComparer','TableNameMaster', EditTableM.Text);
    WriteString('TableDataComparer','TableNameTarget', EditTableT.Text);
    WriteInteger('TableDataComparer','DataCompareMode', cbCompareMode.ItemIndex);
    WriteString('TableDataComparer','CustomKeyFields', EditCustomFlds.Text);
    WriteString('TableDataComparer','ExcludeDataFields', EditExclFlds.Text);
    Free;
  end;
end;

procedure TForm1.BtnSelFileClick(Sender: TObject);
begin
  if (Sender = BtnSelDataBaseMaster) or (Sender = BtnSelDatabaseTarget) then begin
    OpenDialog1.Filter := 'InterBase Databases (*.GDB)|*.gdb|All Files (*.*)|*.*';
  end;
  if (Sender = BtnSelScriptMaster) or (Sender = BtnSelScriptTarget) then begin
    OpenDialog1.Filter := 'SQL Script (*.SQL)|*.sql|SQL Script (*.DDL)|*.ddl|All Files (*.*)|*.*';
  end;
  if OpenDialog1.Execute then begin
    if Sender = BtnSelDatabaseMaster then
      fnDatabaseMaster.Text := OpenDialog1.FileName;
    if Sender = BtnSelDatabaseTarget then
      fnDatabaseTarget.Text := OpenDialog1.FileName;
    if Sender = BtnSelScriptMaster then
      fnScriptMaster.Text := OpenDialog1.FileName;
    if Sender = BtnSelScriptTarget then
      fnScriptTarget.Text := OpenDialog1.FileName;
  end;
end;

procedure TForm1.ExtractMasterDatabase();
var
  ErrFlg: Boolean;
begin
  with DataMd do begin

    ErrFlg := False;
    try
      DBCConnection1.Connected := True;
    except on E: Exception do begin
      ShowMessage(E.Message);
      MemoExtr.Lines.Add('Error: '+E.Message);
      ErrFlg := True;
      UniConnection1.Connected := false;
    end end;
    if ErrFlg then exit;
    MemoExtr.Lines.Add('Database: '+fnDatabaseMaster.Text+' open.');
    DBStructure1.Clear;
    IBDBExtract1.ExtractDatabase;
    DBCConnection1.Connected := False;
  end;
end;

procedure TForm1.ExtractTargetDatabase;
var
  ErrFlg: Boolean;
begin
  with DataMd do begin

    ErrFlg := False;
    try
      DBCConnection2.Connected := True;
    except on E: Exception do begin
      ShowMessage(E.Message);
      MemoExtr.Lines.Add('Error: '+E.Message);
      ErrFlg := True;
      UniConnection2.Connected := false;
    end end;
    if ErrFlg then exit;
    MemoExtr.Lines.Add('Database: '+fnDatabaseTarget.Text+' open.');
    DBStructure2.Clear;
    IBDBExtract2.ExtractDatabase;
    DBCConnection2.Connected := False;
  end;
end;

procedure TForm1.ExtractMasterScript();
begin
  IBScriptExtract1.ScriptFileName := fnScriptMaster.Text;
  IBScriptExtract1.ExtractDatabase;
end;

procedure TForm1.ExtractTargetScript();
begin
  IBScriptExtract2.ScriptFileName := fnScriptTarget.Text;
  IBScriptExtract2.ExtractDatabase;
end;

procedure TForm1.BtnExtractClick(Sender: TObject);
begin
  BtnExtract.Enabled := False;
  if rbTargetFromDatabase.Checked then ExtractTargetDatabase();
  if rbTargetFromScript.Checked then ExtractTargetScript();
  if rbMasterFromDatabase.Checked then ExtractMasterDatabase();
  if rbMasterFromScript.Checked then ExtractMasterScript();
  BtnExtract.Enabled := True;
  MemoExtr.Lines.Add('< Extract finished. >');
end;

procedure TForm1.LogNextLine(Sender: TObject; LogLine: String);
begin
  MemoExtr.Lines.Add(LogLine);
  MemoExtr.Perform(EM_SCROLLCARET, 0, 0);
end;

procedure TForm1.BtnCompareClick(Sender: TObject);
begin
  MemoResult.Lines.Clear;
  BtnCompare.Enabled := False;
  try
    DBComparer1.CompareDatabases();
    MemoResult.Lines.BeginUpdate();
    DBComparer1.SQLExec.GetScript(MemoResult.Lines);
    MemoResult.Lines.EndUpdate();
  except on E: Exception do begin
    ShowMessage(E.Message);
    MemoResult.Lines.Add('Error: '+E.Message);
  end end;
  MemoExtr.Lines.Add('< Compare finished. >');
  BtnCompare.Enabled := True;
end;

procedure TForm1.rbFromClick(Sender: TObject);
begin
  if rbMasterFromDatabase.Checked then begin
    fnDatabaseMaster.Enabled := True;
    BtnSelDatabaseMaster.Enabled := True;
    fnScriptMaster.Enabled := False;
    BtnSelScriptMaster.Enabled := False;
    EditMUser.Enabled := True;
    EditMPswd.Enabled := True;
    LabelMUser.Enabled := True;
    LabelMPswd.Enabled := True;
    fnDatabaseMaster.Color := clWindow;
    fnScriptMaster.Color := clBtnFace;
  end
  else begin
    fnDatabaseMaster.Enabled := False;
    BtnSelDatabaseMaster.Enabled := False;
    fnScriptMaster.Enabled := True;
    BtnSelScriptMaster.Enabled := True;
    EditMUser.Enabled := False;
    EditMPswd.Enabled := False;
    LabelMUser.Enabled := False;
    LabelMPswd.Enabled := False;
    fnDatabaseMaster.Color := clBtnFace;
    fnScriptMaster.Color := clWindow;
  end;
  if rbTargetFromDatabase.Checked then begin
    fnDatabaseTarget.Enabled := True;
    BtnSelDatabaseTarget.Enabled := True;
    fnScriptTarget.Enabled := False;
    BtnSelScriptTarget.Enabled := False;
    EditTUser.Enabled := True;
    EditTPswd.Enabled := True;
    LabelTUser.Enabled := True;
    LabelTPswd.Enabled := True;
    fnDatabaseTarget.Color := clWindow;
    fnScriptTarget.Color := clBtnFace;
  end
  else begin
    fnDatabaseTarget.Enabled := False;
    BtnSelDatabaseTarget.Enabled := False;
    fnScriptTarget.Enabled := True;
    BtnSelScriptTarget.Enabled := True;
    EditTUser.Enabled := False;
    EditTPswd.Enabled := False;
    LabelTUser.Enabled := False;
    LabelTPswd.Enabled := False;
    fnDatabaseTarget.Color := clBtnFace;
    fnScriptTarget.Color := clWindow;
  end;
end;

procedure TForm1.ProgressUpdate(Sender: TObject; Value, MaxValue: Integer);
begin
  Gauge1.Max := MaxValue;
  Gauge1.Position := Value;
end;

procedure TForm1.BtnUpdateClick(Sender: TObject);
begin
  IBSQLExec.ExecuteScript();
end;

procedure TForm1.DecodeFields(aFlds: String; aFields: TStrings);
var
  s: String;
  p: integer;
  FldName: String;

  function CheckUpper(fldName: String): string;
  begin
    if (Length(fldName) > 0) and (fldName[1] in ['A'..'Z','a'..'z']) then
      Result := AnsiUpperCase(fldName)
    else Result := fldName;
  end;

begin
  aFields.Clear;
  s := aFlds;
  p := pos(',', s);
  while p > 0 do begin
    FldName := Trim(Copy(s, 1, p-1));
    if fldName <> '' then begin
      aFields.Add(CheckUpper(FldName));
    end;
    Delete(s, 1, p);
    p := pos(',', s);
  end;
  FldName := Trim(s);
  if FldName <> '' then aFields.Add(CheckUpper(FldName));
end;

procedure TForm1.BtnDataCompareClick(Sender: TObject);
begin
  if (not rbMasterFromDatabase.Checked) or (not rbTargetFromDatabase.Checked) then begin
    ShowMessage('Table data comparer find differences only in database-tables...');
    rbMasterFromDatabase.Checked := True;
    rbTargetFromDatabase.Checked := True;
    rbFromClick(nil);
    exit;
  end;
  TableDataComparer.TableNameMaster := EditTableM.Text;
  TableDataComparer.TableNameTarget := EditTableT.Text;
  DecodeFields(EditCustomFlds.Text, TableDataComparer.CustomKeyFields);
  DecodeFields(EditExclFlds.Text, TableDataComparer.ExcludeDataFields);
  IBSQLExec.Clear;
  MemoResult.Clear;
  if TableDataComparer.CompareData() then begin
    IBSQLExec.GetScript(MemoResult.Lines);
  end;
end;

procedure TForm1.EditTableMChange(Sender: TObject);
begin
  EditTableT.Text := EditTableM.Text;
end;

procedure TForm1.TableDataComparerProgressUpdateMaster(Sender: TObject;
  Value, MaxValue: Integer);
begin
  Gauge1.Max := MaxValue;
  Gauge1.Position := Value;
end;

procedure TForm1.TableDataComparerProgressUpdateTarget(Sender: TObject;
  Value, MaxValue: Integer);
begin
  Gauge2.Max := MaxValue;
  Gauge2.Position := Value;
end;

procedure TForm1.cbCompareModeChange(Sender: TObject);
begin
  EditCustomFlds.Enabled := cbCompareMode.ItemIndex = 3;
  LabelCustomFlds.Enabled := cbCompareMode.ItemIndex = 3;
end;

initialization
  dbcCompObj.AddToLog := AddToExtrLog;
end.

File: CBuilder/DBComparerDemoUniDAC/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
#include "IBDataModule.h"
#include <inifiles.hpp>

//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "dbcClasses"
#pragma link "dbcCustomScriptExtract"
#pragma link "dbcDBComparer"
#pragma link "dbcDBStructure"
#pragma link "dbcIBDatabaseExtract"
#pragma link "dbcIBScriptExtract"
#pragma link "dbcIBSQLExec"
#pragma link "dbcSQL_Exec"
#pragma link "dbcTableDataComparer"
#pragma resource "*.dfm"

TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

UnicodeString __fastcall IniFileName()
{
  return ExtractFilePath(ParamStr(0)) + "CompDemo.Ini";
}

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  fnDatabaseMaster->Text = "";
  fnDatabaseTarget->Text = "";
  fnScriptMaster->Text = "";
  fnScriptTarget->Text = "";
  MemoExtr->Clear();
  MemoResult->Clear();
  
  TIniFile *ini = new TIniFile(IniFileName());
  __try
  {
    fnDatabaseMaster->Text = ini->ReadString("Databases","MasterDB","");
    fnDatabaseTarget->Text = ini->ReadString("Databases","TargetDB","");
    fnScriptMaster->Text = ini->ReadString("Databases","MasterScript","");
    fnScriptTarget->Text = ini->ReadString("Databases","TargetScript","");
    bool MFromD = ini->ReadBool("Databases","MasterFromDatabase",true);
    bool TFromD = ini->ReadBool("Databases","TargetFromDatabase",true);

    if (MFromD)
      rbMasterFromDatabase->Checked = true;
    else
      rbMasterFromScript->Checked = true;
    if (TFromD)
      rbTargetFromDatabase->Checked = true;
    else
      rbTargetFromScript->Checked = true;

    EditTableM->Text = ini->ReadString("TableDataComparer","TableNameMaster", "");
    EditTableT->Text = ini->ReadString("TableDataComparer","TableNameTarget", "");
    cbCompareMode->ItemIndex = ini->ReadInteger("TableDataComparer","DataCompareMode", 0);
    cbCompareModeChange(cbCompareMode);
    EditCustomFlds->Text = ini->ReadString("TableDataComparer","CustomKeyFields", "");
    EditExclFlds->Text = ini->ReadString("TableDataComparer","ExcludeDataFields", "");
  }
  __finally
  {
    delete ini;
  }
  rbFromClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnSelFileClick(TObject *Sender)
{
  if ((Sender == BtnSelDatabaseMaster) || (Sender == BtnSelDatabaseTarget))
  {
    OpenDialog1->Filter = "InterBase Databases (*.GDB)|*.gdb|All Files (*.*)|*.*";
  }
  if ((Sender == BtnSelScriptMaster) || (Sender == BtnSelScriptTarget))
  {
    OpenDialog1->Filter = "SQL Script (*.SQL)|*.sql|SQL Script (*.DDL)|*.ddl|All Files (*.*)|*.*";
  }
  if (OpenDialog1->Execute())
  {
    if (Sender == BtnSelDatabaseMaster)
      fnDatabaseMaster->Text = OpenDialog1->FileName;
    if (Sender == BtnSelDatabaseTarget)
      fnDatabaseTarget->Text = OpenDialog1->FileName;
    if (Sender == BtnSelScriptMaster)
      fnScriptMaster->Text = OpenDialog1->FileName;
    if (Sender == BtnSelScriptTarget)
      fnScriptTarget->Text = OpenDialog1->FileName;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnExtractClick(TObject *Sender)
{
  BtnExtract->Enabled = false;
  if (rbTargetFromDatabase->Checked)
    ExtractTargetDatabase();
  if (rbTargetFromScript->Checked)
    ExtractTargetScript();
  if (rbMasterFromDatabase->Checked)
    ExtractMasterDatabase();
  if (rbMasterFromScript->Checked)
    ExtractMasterScript();
  BtnExtract->Enabled = true;
  MemoExtr->Lines->Add("< Extract finished. >");
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormDestroy(TObject *Sender)
{
  TIniFile *ini = new TIniFile(IniFileName());
  __try
  {
    ini->WriteString("Databases", "MasterDB", fnDatabaseMaster->Text);
    ini->WriteString("Databases", "TargetDB", fnDatabaseTarget->Text);
    ini->WriteString("Databases", "MasterScript", fnScriptMaster->Text);
    ini->WriteString("Databases", "TargetScript", fnScriptTarget->Text);
    ini->WriteBool("Databases","MasterFromDatabase",rbMasterFromDatabase->Checked);
    ini->WriteBool("Databases","TargetFromDatabase",rbTargetFromDatabase->Checked);
    ini->WriteString("TableDataComparer","TableNameMaster", EditTableM->Text);
    ini->WriteString("TableDataComparer","TableNameTarget", EditTableT->Text);
    ini->WriteInteger("TableDataComparer","DataCompareMode", cbCompareMode->ItemIndex);
    ini->WriteString("TableDataComparer","CustomKeyFields", EditCustomFlds->Text);
    ini->WriteString("TableDataComparer","ExcludeDataFields", EditExclFlds->Text);
  }
  __finally
  {
    delete ini;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::LogNextLine(TObject *Sender, UnicodeString LogLine)
{
  MemoExtr->Lines->Add(LogLine);
  MemoExtr->Perform(EM_SCROLLCARET, 0, 0);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnCompareClick(TObject *Sender)
{
  MemoResult->Lines->Clear();
  BtnCompare->Enabled = false;
  try
  {
    DBComparer1->CompareDatabases();
    MemoResult->Lines->BeginUpdate();
    DBComparer1->SQLExec->GetScript(MemoResult->Lines);
    MemoResult->Lines->EndUpdate();
  }
  catch (Exception &E)
  {
    Dialogs::ShowMessage(E.Message);
    MemoResult->Lines->Add("Error: " + E.Message);
  }
  MemoExtr->Lines->Add("< Compare finished. >");
  BtnCompare->Enabled = true;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::rbFromClick(TObject *Sender)
{
  if (rbMasterFromDatabase->Checked)
  {
    fnDatabaseMaster->Enabled = true;
    BtnSelDatabaseMaster->Enabled = true;
    fnScriptMaster->Enabled = false;
    BtnSelScriptMaster->Enabled = false;
    EditMUser->Enabled = true;
    EditMPswd->Enabled = true;
    LabelMUser->Enabled = true;
    LabelMPswd->Enabled = true;
    fnDatabaseMaster->Color = clWindow;
    fnScriptMaster->Color = clBtnFace;
  } else
  {
    fnDatabaseMaster->Enabled = false;
    BtnSelDatabaseMaster->Enabled = false;
    fnScriptMaster->Enabled = true;
    BtnSelScriptMaster->Enabled = true;
    EditMUser->Enabled = false;
    EditMPswd->Enabled = false;
    LabelMUser->Enabled = false;
    LabelMPswd->Enabled = false;
    fnDatabaseMaster->Color = clBtnFace;
    fnScriptMaster->Color = clWindow;
  }
  if (rbTargetFromDatabase->Checked)
  {
    fnDatabaseTarget->Enabled = true;
    BtnSelDatabaseTarget->Enabled = true;
    fnScriptTarget->Enabled = false;
    BtnSelScriptTarget->Enabled = false;
    EditTUser->Enabled = true;
    EditTPswd->Enabled = true;
    LabelTUser->Enabled = true;
    LabelTPswd->Enabled = true;
    fnDatabaseTarget->Color = clWindow;
    fnScriptTarget->Color = clBtnFace;
  } else
  {
    fnDatabaseTarget->Enabled = false;
    BtnSelDatabaseTarget->Enabled = false;
    fnScriptTarget->Enabled = true;
    BtnSelScriptTarget->Enabled = true;
    EditTUser->Enabled = false;
    EditTPswd->Enabled = false;
    LabelTUser->Enabled = false;
    LabelTPswd->Enabled = false;
    fnDatabaseTarget->Color = clBtnFace;
    fnScriptTarget->Color = clWindow;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ProgressUpdate(TObject *Sender, int Value, int MaxValue)
{
  Gauge1->Max = MaxValue;
  Gauge1->Position = Value;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnUpdateClick(TObject *Sender)
{
  IBSQLExec->ExecuteScript();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::BtnDataCompareClick(TObject *Sender)
{
  if ((!rbMasterFromDatabase->Checked) || (!rbTargetFromDatabase->Checked))
  {
    Dialogs::ShowMessage("Table data comparer find differences only in database-tables...");
    rbMasterFromDatabase->Checked = true;
    rbTargetFromDatabase->Checked = true;
    rbFromClick(NULL);
    return;
  }
  
  TableDataComparer->TableNameMaster = EditTableM->Text;
  TableDataComparer->TableNameTarget = EditTableT->Text;
  DecodeFields(EditCustomFlds->Text, TableDataComparer->CustomKeyFields);
  DecodeFields(EditExclFlds->Text, TableDataComparer->ExcludeDataFields);
  IBSQLExec->Clear();
  MemoResult->Clear();
  
  if (TableDataComparer->CompareData())
  {
    IBSQLExec->GetScript(MemoResult->Lines);
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm1::EditTableMChange(TObject *Sender)
{
  EditTableT->Text = EditTableM->Text;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::TableDataComparerProgressUpdateMaster(TObject *Sender, int Value, int MaxValue)
{
  Gauge1->Max = MaxValue;
  Gauge1->Position = Value;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::TableDataComparerProgressUpdateTarget(TObject *Sender, int Value, int MaxValue)
{
  Gauge2->Max = MaxValue;
  Gauge2->Position = Value;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::cbCompareModeChange(TObject *Sender)
{
  EditCustomFlds->Enabled = cbCompareMode->ItemIndex == 3;
  LabelCustomFlds->Enabled = cbCompareMode->ItemIndex == 3;
}
//---------------------------------------------------------------------------

UnicodeString __fastcall CheckUpper(UnicodeString fldName)
{
  if ((fldName.Length() > 0) && (
    ((fldName[1] >= 'a') && (fldName[1] <= 'z'))
    || ((fldName[1] >= 'A') && (fldName[1] <= 'Z'))))
  {
    return AnsiUpperCase(fldName);
  }
  return fldName;
}

void __fastcall TForm1::DecodeFields(UnicodeString aFlds, TStrings *aFields)
{
  aFields->Clear();
  UnicodeString s = aFlds;
  UnicodeString FldName;

  int p = s.Pos(",");
  while (p > 0)
  {
    FldName = s.SubString(1, p - 1).Trim();
    if (FldName != "")
      aFields->Add(CheckUpper(FldName));

    s.Delete(1, p);
    p = s.Pos(",");
  }
  
  FldName = s.Trim();
  if (FldName != "")
    aFields->Add(CheckUpper(FldName));
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractMasterDatabase()
{
  bool ErrFlg = false;
  try
  {
    DataMd->DBCConnection1->Connected = true;
  }
  catch (Exception &E)
  {
    Dialogs::ShowMessage(E.Message);
    MemoExtr->Lines->Add("Error: " + E.Message);
    ErrFlg = true;
	DataMd->UniConnection1->Connected = false;
  }
  if (ErrFlg) return;

  MemoExtr->Lines->Add("Database: " + fnDatabaseMaster->Text + " open.");
  DBStructure1->Clear();
  IBDBExtract1->ExtractDatabase();
  DataMd->DBCConnection1->Connected = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractTargetDatabase()
{
  bool ErrFlg = false;
  try
  {
	DataMd->DBCConnection2->Connected = true;
  }
  catch (Exception &E)
  {
	Dialogs::ShowMessage(E.Message);
	MemoExtr->Lines->Add("Error: " + E.Message);
	ErrFlg = true;
	DataMd->UniConnection2->Connected = false;
  }
  if (ErrFlg) return;

  MemoExtr->Lines->Add("Database: " + fnDatabaseTarget->Text + " open.");
  DBStructure2->Clear();
  IBDBExtract2->ExtractDatabase();
  DataMd->DBCConnection2->Connected = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractMasterScript()
{
  IBScriptExtract1->ScriptFileName = fnScriptMaster->Text;
  IBScriptExtract1->ExtractDatabase();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ExtractTargetScript()
{
  IBScriptExtract2->ScriptFileName = fnScriptTarget->Text;
  IBScriptExtract2->ExtractDatabase();
}
//---------------------------------------------------------------------------



